Autosave: Do not use autosave for drafts. Minor refactor.

aerych opened the pull request:
Fixes #13079 
This PR addresses a situation where previewing changes made to a draft post would result in comments for that post being incorrectly turned off.  This is a quirk of the v1.1 autosave endpoint and we have opted to address the issue in the apps rather than in the API.  See p3hLNG-15Z-p2 for more background on that decision. 

I've added a check to see if the post has a draft status when autosaving. If the post or the original post is a draft, we query the API for the post status on the server.  This is an extra safety measure for the edge case where the post has been published outside of the app, but the app still has the draft.

I've also performed a minor refactor to flatten some conditionals and shorten methods. The autosave method was getting rather large with lots of nesting. 

My Objective C is a bit rusty (yay Swift!) and [block syntax](http://fuckingblocksyntax.com/) remains a right pita. Please give special attention to the block factories and typedefs when reviewing.

@shiki could I trouble you for a review of this one? 

cc @oguzkocer 

To test:
- Use the develop branch to confirm the bug by following the steps outlined in [this comment](https://github.com/wordpress-mobile/WordPress-iOS/issues/13079#issuecomment-580982420) in the issue.
- Switch to this branch and repeat the steps to confirm the bug is resolved. 
- Confirm that the autosave behavior is not altered for posts with other statuses. 

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

shiki commented on WordPress/Classes/Services/PostService.m:
```diff
@@ -333,70 +337,129 @@ - (void)autoSave:(AbstractPost *)post
+                [self uploadPost:post
+                success:^(AbstractPost * _Nonnull post) {
+                    success(post, nil);
+                } failure:failure];
```
🛑 This can cause a Preview to **publish** a post. I was able to do that using these steps:

1. Create a draft and upload it. 
2. Edit the draft.
3. Set Post Settings → Status to Published
4. Go back to Editor.
5. More → Preview

This block of code will be hit and the post will be uploaded with a `.published` status. 

Luckily, we could probably use the `forceDraftIfCreating:YES` argument in here. We used that to solve problems like this.

aerych replied:
Oh! Good catch!

aerych replied:
Looks like `forceDraftIfCreating` checks for a non-zero postID and only forces a draft if zero.  Not too keen on modifying the behavior there so I'll address this case a different way.

shiki commented on WordPress/Classes/Services/PostService.m:
```diff
@@ -333,70 +337,129 @@ - (void)autoSave:(AbstractPost *)post
+    NSManagedObjectID *postObjectID = post.objectID;
+    RemotePost *remotePost = [self remotePostWithPost:post];
+
+    AutosaveSuccessBlock successBlock = [self wrappedAutosaveSuccessBlock:postObjectID success:success];
```
In line with my comment below about the `failureBlock` name, we could also rename this to `autosaveSuccessBlock`. It's not a great name but it'll probably make it less surprising why we use `success` and `successBlock` sometimes. 

Admittedly, I didn't see this ambiguous stuff when I last touched this method. 😄

shiki commented on WordPress/Classes/Services/PostService.m:
```diff
@@ -333,70 +337,129 @@ - (void)autoSave:(AbstractPost *)post
+- (AutosaveSuccessBlock)wrappedAutosaveSuccessBlock:(NSManagedObjectID *)postObjectID
+                                            success:(void (^)(AbstractPost *post, NSString *previewURL))success
+{
+    return ^(RemotePost *post, NSString *previewURL) {
```
It looks like we never actually used `post` in the old code. 😅 Probably not worth it to change right now though.

shiki commented on WordPress/Classes/Services/PostService.m:
```diff
@@ -312,18 +312,22 @@ - (void)uploadPost:(AbstractPost *)post
+                           userInfo:@{ NSLocalizedDescriptionKey : @"Previews are unavailable for this kind of post." }];
+}
+
+- (AutosaveFailureBlock)wrappedAutosaveFailureBlock:(AbstractPost *)post failure:(void (^)(NSError * _Nullable error))failure
```
We could probably pass `postObjectID` instead so this is more consistent with the `wrappedAutosaveSuccessBlock`'s signature.

aerych replied:
I considered that. I opted to keep the type AbstractPost since that's what we ultimately need to obtain. NSManagedObjectID wouldn't be a safe type in this case, strictly speaking, and so seems less future proof.

aerych commented:
> We also created a lot of tests for PostService in PostServiceWPComTests. Do you think it's worth it to add a test for this PR?

Thanks for the nudge.  Most of the functionality is the same, just moved around. I've added a test to ensure the correct endpoint is called for drafts. 

> The fewer methods there are in PostService, the easier it is to grok

True facts, but yeah, let's save it for another PR :) 

I've made changes and left a reply or two.  Ready for another round!

shiki commented on WordPress/WordPressTest/Services/PostServiceWPComTests.swift:
```diff
@@ -154,6 +154,27 @@ class PostServiceWPComTests: XCTestCase {
         expect(post.remoteStatus).to(equal(.sync))
     }
 
+    func testAutoSavingADraftWillCallTheUpdateEndpointInstead() {
```
Perfect! 🎉

shiki commented on WordPress/Classes/Services/PostService.m:
```diff
@@ -333,70 +337,137 @@ - (void)autoSave:(AbstractPost *)post
+        return;
+    }
+
+    if ([post originalIsDraft]) {
```
⚠️ I'm sorry I didn't see this before. I found another scenario where the comments can still be disabled. It happens in the offline posting cancellation flow:

1. Create a draft and upload it. 
2. Go offline. 
3. Still in the Post List, tap More → Publish Now. The upload should fail and there should be a Cancel button visible. 
4. Tap on Cancel button. 
5. Go online.

Going online will trigger auto-uploading. However, the **explicit cancelation** of the user will disallow "uploads". In cases like this, we still try to save the user's changes by autosaving instead. This is where that decision happens:

https://github.com/wordpress-mobile/WordPress-iOS/blob/c957f3d86b4a4481ca8e5cc3b0c3e20473a4f810/WordPress/Classes/Services/PostAutoUploadInteractor.swift#L37-L63

So, eventually, `PostService.handleAutoSaveWithRestRemote` will still be called. The problem is the **`AbstractPost.status` will still be `.publish`** at this point. This `[post originalIsDraft]` condition will **fail** and the routine will move to line 463 which causes the comment-disabling bug again:

https://github.com/wordpress-mobile/WordPress-iOS/blob/894bd2cf8c8477e1c963a0dfaeb94152a189d248/WordPress/Classes/Services/PostService.m#L462-L465

I think we can get rid of this `[post originalIsDraft]` and condition and always attempt to do a `getPostWithID` before calling `autosave`. And also get rid of the `autosave` call in line 463.

aerych replied:
Good catch! The tweak makes sense and it seems the safest thing to always check the current remote value of status.

oguzkocer commented:
Hey @aerych & @shiki! I haven't gone through the PR, but I just wanted to ask if you considered what happens if the remote status of the post is draft and the local status is not. I imagine there are some differences in iOS and Android, so I don't know how or if this would affect things in iOS. On Android, if we don't handle this, we'd end up overwriting the local status of the post so I figured maybe something like that is also happening in iOS.

shiki commented on WordPress/Classes/Services/PostService.m:
```diff
@@ -333,70 +337,137 @@ - (void)autoSave:(AbstractPost *)post
+                success:^(AbstractPost * _Nonnull post) {
+                    post.status = savedStatus;
+                    success(post, nil);
+                } failure:failure];
```
Oh, I think we should also restore the `savedStatus` when the `failure` block is called.

aerych replied:
Doh!  Right you are.

shiki commented:
@oguzkocer Yes. We (or Eric did) added a routine that before calling POST to upload, the `status` will be reset to `draft`. 

https://github.com/wordpress-mobile/WordPress-iOS/blob/894bd2cf8c8477e1c963a0dfaeb94152a189d248/WordPress/Classes/Services/PostService.m#L437-L448

aerych commented:
Thanks again @shiki, and thanks for the thorough description of the other trigger. It should be addressed now. Ready for another peek.

aerych commented:
Thanks again @shiki !
